src/pcapi/core/offers/api.py

```diff
@@ -260,37 +285,100 @@ def as_utc_without_timezone(d: datetime.datetime) -> datetime.datetime:
+    bookings = bookings_repository.find_not_cancelled_bookings_by_stock(stock)
+    if bookings:
+        bookings = update_confirmation_dates(bookings, stock.beginningDatetime)
+        date_in_two_days = datetime.datetime.utcnow() + datetime.timedelta(days=2)
+        check_event_is_in_more_than_48_hours = stock.beginningDatetime > date_in_two_days
+        if check_event_is_in_more_than_48_hours:
+            bookings = _invalidate_bookings(bookings)
+        try:
+            user_emails.send_batch_stock_postponement_emails_to_users(bookings, send_email=mailing.send_raw_email)
+        except mailing.MailServiceException as exc:
+            # fmt: off
+            app.logger.exception(
+                "Could not notify beneficiaries about update of stock=%s: %s",
+                stock.id,
+                exc,
+            )
+            # fmt: on
+
+
+def edit_stock(
+    stock: Stock,
+    price: float = None,
+    quantity: int = None,
+    beginning: datetime.datetime = None,
+    booking_limit_datetime: datetime.datetime = None,
+) -> Stock:
+    previous_beginning = stock.beginningDatetime
+    stock = _edit_stock(
+        stock,
+        price,
+        quantity,
+        beginning,
+        booking_limit_datetime,
+    )
+
     repository.save(stock)
 
-    if beginning != previous_beginning:
-        bookings = bookings_repository.find_not_cancelled_bookings_by_stock(stock)
-        if bookings:
-            bookings = update_confirmation_dates(bookings, beginning)
-            date_in_two_days = datetime.datetime.utcnow() + datetime.timedelta(days=2)
-            check_event_is_in_more_than_48_hours = beginning > date_in_two_days
-            if check_event_is_in_more_than_48_hours:
-                bookings = _invalidate_bookings(bookings)
-            try:
-                user_emails.send_batch_stock_postponement_emails_to_users(bookings, send_email=mailing.send_raw_email)
-            except mailing.MailServiceException as exc:
-                # fmt: off
-                app.logger.exception(
-                    "Could not notify beneficiaries about update of stock=%s: %s",
-                    stock.id,
-                    exc,
-                )
-                # fmt: on
+    if stock.beginningDatetime != previous_beginning:
+        _notify_upon_stock_edit(stock)
 
     if feature_queries.is_active(FeatureToggle.SYNCHRONIZE_ALGOLIA):
         redis.add_offer_id(client=app.redis_client, offer_id=stock.offerId)
-
     return stock
 
 
+def upsert_stocks(offer_id: int, stock_data_list: List[Union[StockCreationBodyModel, StockEditionBodyModel]]):
+    stocks = []
+    edited_stocks = []
+    edited_stocks_previous_beginnings = {}
+
+    offer = offer_queries.get_offer_by_id(offer_id)
+
+    for idx, data in enumerate(stock_data_list):
+        if isinstance(data, StockEditionBodyModel):
+            stock = Stock.queryNotSoftDeleted().filter_by(id=data.id).join(Offer, VenueSQLEntity).first_or_404()
```

Comment:
Il faudrait ici s'assurer qu'on reçoit bien un stock_id correspondant à l'`offer_id` fournie plus haut. :

```suggestion
            stock = Stock.queryNotSoftDeleted().filter_by(offerId=offer_id, id=data.id).join(Offer, VenueSQLEntity).first_or_404()
```

Sinon, un utilisateur pourrait passer un `offer_id` qui lui appartient mais des `stock_id` d'autres offres ne lui appartenant pas.

Reply:
Ah, en effet...
Par contre plutôt une 400 ou 503 en mode "vous n'avez pas le droit de modifier ce stock" non ?

D'ailleurs il faut enlever les `.join(Offer, VenueSQLEntity)`, ça doit être un legacy d'autre chose